builtin/logical/pki/crl_util.go

```diff
@@ -219,11 +351,106 @@ func buildCRL(ctx context.Context, b *backend, req *logical.Request, forceNew bo
 		} else {
 			newRevCert.RevocationTime = time.Unix(revInfo.RevocationTime, 0).UTC()
 		}
-		revokedCerts = append(revokedCerts, newRevCert)
+
+		// If we have a CertificateIssuer field on the revocation entry,
+		// prefer it to manually checking each issuer signature, assuming it
+		// appears valid. Its highly unlikely for two different issuers
+		// to have the same id (after the first was deleted).
+		if len(revInfo.CertificateIssuer) > 0 {
+			issuerId := revInfo.CertificateIssuer
+			if _, issuerExists := issuerIDCertMap[issuerId]; issuerExists {
+				revokedCertsMap[issuerId] = append(revokedCertsMap[issuerId], newRevCert)
+				continue
+			}
+
+			// Otherwise, fall through and update the entry.
+		}
+
+		// Now we need to assign the revoked certificate to an issuer.
+		foundParent := false
+		for issuerId, issuerCert := range issuerIDCertMap {
+			if bytes.Equal(revokedCert.RawIssuer, issuerCert.RawSubject) {
+				if err := revokedCert.CheckSignatureFrom(issuerCert); err == nil {
+					// Valid mapping. Add it to the specified entry.
+					revokedCertsMap[issuerId] = append(revokedCertsMap[issuerId], newRevCert)
+					revInfo.CertificateIssuer = issuerId
+					foundParent = true
+					break
+				}
+			}
+		}
+
+		if !foundParent {
+			// If the parent isn't found, add it to the unassigned bucket.
+			unassignedCerts = append(unassignedCerts, newRevCert)
+		} else {
+			// When the CertificateIssuer field wasn't found on the existing
+			// entry (or was invalid), and we've found a new value for it,
+			// we should update the entry to make future CRL builds faster.
+			revokedEntry, err = logical.StorageEntryJSON("revoked/"+normalizeSerial(serial), revInfo)
```

Comment:
So our call to normalizeSerial here is a bit touchy imho. 

We list the revoked/ storage above and fetch serial storage as is (which makes sense), with this call to normalizeSerial there is a chance that we write out the same serial within a different entry than what we loaded.

Reply:
Also please consider using the constant `revokedPath`

Reply:
Ah, I see. Yeah, I was looking at the code in `revokeCert`, which writes at the correct serial location. I've fixed it :-)